core/spoofax.eclipse/src/main/java/mb/spoofax/eclipse/editor/SpoofaxHyperlinkDetector.java

```
@@ -0,0 +1,76 @@
+import org.eclipse.jface.text.hyperlink.IHyperlinkDetector;
+
+import java.util.Collections;
+import java.util.Objects;
+
+public class SpoofaxHyperlinkDetector implements IHyperlinkDetector {
+    private final SpoofaxEditorBase editorBase;
+    private final LanguageComponent languageComponent;
+    private final PieComponent pieComponent;
+
+    public SpoofaxHyperlinkDetector(SpoofaxEditorBase editorBase, LanguageComponent languageComponent, PieComponent pieComponent) {
+        this.editorBase = editorBase;
+        this.languageComponent = languageComponent;
+        this.pieComponent = pieComponent;
+    }
+
+    @Override
+    public @Nullable IHyperlink[] detectHyperlinks(ITextViewer textViewer, IRegion region, boolean canShowMultipleHyperlinks) {
+        final ResourceKey file = new EclipseResourcePath(editorBase.file);
+        final @Nullable ResourcePath rootDirectory = editorBase.project != null ? new EclipseResourcePath(editorBase.project) : null;
+        final Region targetRegion = Region.fromOffsetLength(region.getOffset(), region.getLength());
+
+        // We cannot do reference resolution if we're not in a project.
+        if(rootDirectory == null) {
+            return null;
+        }
+
+        try(final MixedSession mixedSession = pieComponent.newSession()) {
+            final TopDownSession topDownSession = mixedSession.updateAffectedBy(Collections.emptySet(), Collections.singleton(Interactivity.Interactive));
+
+            final Option<ReferenceResolutionResult> resolveResult = topDownSession.requireWithoutObserving(
+                languageComponent.getLanguageInstance().createResolveTask(rootDirectory, file, targetRegion)
+            );
+
+            if(!resolveResult.isSome()) {
+                return null;
+            }
+
+            return referenceResolutionResultToHyperlinks(resolveResult.get());
+        } catch(ExecException | InterruptedException e) {
+            return null;
```

Please log this exception as an error, silently failing is bad.

---

Should I also handle the `InterruptedException`?

---

`InterruptedException` should be handled as it is now, just immediately return null.

---

Updated.